contracts/bonding/BondingManager.sol

```diff
@@ -350,142 +309,231 @@ contract BondingManager is IBondingManager, Controllable {
-     * @param _transcoder Address of transcoder
+     * @dev Slash a transcoder. Slashing can be invoked by the protocol or a finder.
+     * @param _transcoder Transcoder address
+     * @param _finder Finder that proved a transcoder violated a slashing condition. Null address if there is no finder
+     * @param _slashAmount Percentage of transcoder bond to be slashed
+     * @param _finderFee Percentage of penalty awarded to finder. Zero if there is no finder
      */
-    function validRewardCall(address _transcoder) public constant returns (bool) {
-        // Check if transcoder is in active set
-        if (!isActiveTranscoder[_transcoder]) return false;
-        // Check if already called for current cycle
-        if (transcoders[_transcoder].rewardRound == roundsManager().currentRound() && transcoders[_transcoder].rewardCycle == roundsManager().cycleNum()) return false;
+    function slashTranscoder(
+        address _transcoder,
+        address _finder,
+        uint64 _slashAmount,
+        uint64 _finderFee
+    )
+        external
+        whenSystemNotPaused
+        onlyJobsManager
+        returns (bool)
+    {
+        // Transcoder must be valid
+        require(transcoderStatus(_transcoder) == TranscoderStatus.Registered);
+
+        Transcoder storage t = transcoders[_transcoder];
+
+        uint256 penalty = t.bondedAmount.mul(_slashAmount).div(100);
+
+        if (penalty > t.bondedAmount) {
+            // Decrease transcoder's total stake by transcoder's bond
+            transcoderPools.decreaseTranscoderStake(_transcoder, t.bondedAmount);
+            // Set transcoder's bond to 0 since
+            // the penalty is greater than its bond
+            t.bondedAmount = 0;
+        } else {
+            // Decrease transcoder's total stake by the penalty
+            transcoderPools.decreaseTranscoderStake(_transcoder, penalty);
+            // Decrease transcoder's bond
+            t.bondedAmount = t.bondedAmount.sub(penalty);
+        }
+
+        // Set withdraw round for delegators
+        transcoders[msg.sender].withdrawRound = roundsManager().currentRound().add(unbondingPeriod);
+
+        if (transcoderPools.isInPools(msg.sender)) {
+            // Remove transcoder from pools
+            transcoderPools.removeTranscoder(msg.sender);
+        }
+
+        if (_finder != address(0)) {
+            // Award finder fee
+            token.transfer(_finder, penalty.mul(_finderFee).div(100));
+        }
+
+        return true;
+    }
+
+    /*
+     * @dev Pseudorandomly elect a currently active transcoder that charges a price per segment less than or equal to the max price per segment for a job
+     * Returns address of elected active transcoder and its price per segment
+     * @param _maxPricePerSegment Max price (in LPT base units) per segment of a stream
+     */
+    function electActiveTranscoder(uint256 _maxPricePerSegment) external constant returns (address, uint256) {
```

Comment:
Let's have a future discussion about charging the transcoder fee or the max price per segment submitted by the broadcaster.